src/duplicacy_utils.go

```diff
@@ -325,24 +340,25 @@ func MatchPath(filePath string, patterns [] string) (included bool) {
              if matchPattern(filePath, pattern[1:]) {
                 return false
             }
-        } else if strings.HasPrefix(pattern, "i:") {
-            matched, err := regexp.MatchString(pattern[2:], filePath)
-            if err != nil {
-                LOG_ERROR("SNAPSHOT_MATCH", "Error during regexp match: %s - %v", pattern, err)
-            }
-            if matched {
-                LOG_TRACE("SNAPSHOT_MATCH", "Regex include comparison for filePath=\"%s\", pattern=\"%s\", matched=%t", filePath, pattern[2:], matched)
-                return true
-            }
-        } else if strings.HasPrefix(pattern, "e:") {
-            allIncludes = false
-            matched, err := regexp.MatchString(pattern[2:], filePath)
-            if err != nil {
-                LOG_ERROR("SNAPSHOT_MATCH", "Error during regexp match: %s - %v", pattern, err)
+        } else if strings.HasPrefix(pattern, "i:") || strings.HasPrefix(pattern, "e:") {
+            if re, found = RegularExpressions[pattern[2:]]; found {
+                LOG_DEBUG("REGEX_FOUND", "Compiled regex for pattern \"%s\" found", pattern)
+                matched = re.MatchString(filePath)
+            } else {
+                LOG_DEBUG("REGEX_NOTFOUND", "Compiled regex for pattern \"%s\" not found!", pattern)
+                matched, _ = regexp.MatchString(pattern[2:], filePath)
```

Comment:
Why wouldn't we simply compile and store the regex, then use it? We should be caching regexes in all cases, not just some.

Reply:
I see your point, but again, I was thinking defensively how to ensure that I'd handle different scenarios. By the time MatchPath is called, all of the filters should have been compiled and stored, by calling IsValidRegex and that is indeed the case as of this PR.

In future updates, if code is introduced that doesn't call IsValidRegex first, then the code here would handle either these where we have already compiled a regex or if for some as of now unknown reason we don't have a compiled regex, it's compiled and stored, etc.

Reply:
You're right that handling the case where the regex isn't in the cache is defensive programming, but what I'm recommending is defensive programming at the next level up: ensuring that you not only do something, but do the right something, if you hit that condition.

You're baking logic for your calling algorithms into these low-level methods, which couples them more tightly than they should be and hinders the ability to reuse them if the need arose. Why should you treat this particular regex any different than any others, just because someone happens not to have called IsValidRegex() first (or just because all the code that's been written so far calls IsValidRegex() first)? And why should you bake in inconsistent behavior now just because you know it doesn't currently matter either way, when it takes almost no more effort to do it the right way and then know that future maintainers won't have to worry about it if they implement something that doesn't call IsValidRegex() first?

Reply:
Changed my mind and have reworked this to compile, store, and used the compiled regex if not already found in the map.

Reply:
Sounds great. It appears our comments were ships passing in the night, but I'm glad we both ended up at the same place.